streamer/m3u8_concater.py

```diff
@@ -0,0 +1,955 @@
+          # for this codec-language-channel triad.
+          (stream_info['BANDWIDTH'],
+           stream_info['AVERAGE-BANDWIDTH']) = MediaPlaylist._get_bandwidth(
+                var_playlists,
+                durations)
+          # Get the codecs from the associated stream variant playlists.
+          stream_info['CODECS'] = MediaPlaylist._get_codec(var_playlists)
+          concat_var_playlist = MediaPlaylist(stream_info)
+          # Set the target duration.
+          concat_aud_playlist.target_duration = MediaPlaylist._max_targer_dur(
+              aud_playlists)
+          for aud_playlist in aud_playlists:
+            concat_aud_playlist.content += aud_playlist.content
+            # Add a discontinuity after each period.
+            concat_aud_playlist.content += '#EXT-X-DISCONTINUITY\n\n'
+          # The audio and the stream variant playlist will be exactly the same.
+          concat_var_playlist.target_duration = concat_aud_playlist.target_duration
+          concat_var_playlist.content = concat_aud_playlist.content
+          concat_aud_only_playlists.extend(
+              [concat_aud_playlist, concat_var_playlist])
+    
+    return concat_aud_only_playlists
+  
+  @staticmethod
+  def concat_vid(all_vid_playlists: List[List['MediaPlaylist']],
+                 durations: List[float]) -> List['MediaPlaylist']:
+    """Concatenates multiple video playlists into one multi-period playlist
+    for many resolutions and codecs, it matches the codecs first, then matches
+    the resolutions.
+    
+    It will pick the closest lower resolution for some period(input) if a high
+    enough resolution was not available.
+    
+    A video playlist is a stream variant playlist based on the Packager's output.
+    """
+    
+    # Get all possible video codecs.  We should not use the all codecs available
+    # in the pipeline config, because for some codecs we might have given the
+    # Packager the 'dash_only' flag.
+    codecs: Set[VideoCodec] = set()
+    # Also get all the available resolutions.
+    resolutions: Set[VideoResolution] = set()
+    for vid_playlists in all_vid_playlists:
+      for vid_playlist in vid_playlists:
+        assert isinstance(vid_playlist.codec, VideoCodec)
+        codecs.add(vid_playlist.codec)
+        resolutions.add(vid_playlist.resolution)
+    
+    # Create and initialize a division map.
+    division: Dict[VideoCodec,
+                   Dict[VideoResolution,
+                        List[MediaPlaylist]]] = {}
+    for codec in codecs:
+      division[codec] = {}
+      for resolution in resolutions:
+        division[codec][resolution] = []
+    
+    # In each period do the following:
+    for vid_playlists in all_vid_playlists:
+      # Initialize a mapping between video codecs and a list of resolutions available.
+      codec_division: Dict[VideoCodec, List[MediaPlaylist]] = {}
+      for codec in codecs:
+        codec_division[codec] = []
```

Comment:
No need to change this in this PR, but just as a tip, you might consider if [collections.defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict) might be a useful tool for this pattern.